### ISSUE TITLE
Allow accessing object properties in conditions

### ISSUE THREAD

**Issue by Mikilll94** [microsoft/TypeScript#57046]
### 🔍 Search Terms

https://www.typescriptlang.org/play?#code/C4TwDgpgBAyg9gWwgFXNAvFA3lMAnOMARgC4oBnYPASwDsBzKAXygB9tcCwAmMymhswDcAKBEBjOLUoVEEAGoBDPGXhJUkKJhz5CpKACIIAD0UIwAGwgHmUReVhyNEUdQBmUABTk5SvADpdYgBKbBEoCKhJaTgrfws4em9fZUCuImCRFggLcmh3Lx8kPzTCblCscMjonziEpKKFVKDy0SYRIA

As you can see in the example above, I am getting errors when accessing the property `prop1` or `prop2`. I don't understand why I am getting this error. The code is safe because I am checking for property existance in `if` condition so TypeScript should be happy.

### ✅ Viability Checklist

- [X] This wouldn't be a breaking change in existing TypeScript/JavaScript code
- [X] This wouldn't change the runtime behavior of existing JavaScript code
- [X] This could be implemented without emitting different JS based on the types of the expressions
- [X] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
- [X] This isn't a request to add a new utility type: https://github.com/microsoft/TypeScript/wiki/No-New-Utility-Types
- [X] This feature would agree with the rest of our Design Goals: https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals

### ⭐ Suggestion

TypeScript should not throw errors when checking the property existance in condition.

### 📃 Motivating Example

I have an API which can return different object structures based on some conditions. I would like to do an `if` check for property existance to select the correct structure. Right now that's not possible because TypeScript doesn't even allow to check for property existance in condition.

### 💻 Use Cases

- Check for property existance in a safe way, i.e. `if (myObj.prop1?.prop2?.prop3)` without getting errors from TypeScript

**Comment by MartinJohns** [comment]
Essentially a duplicate of #42775.

Objects in TypeScript are not sealed. Additional properties can exist. The object `{prop2: 'abc', prop1: 123 }` is assignable to your union type, but checking for the existence of `prop1` does **not** mean it's a `string`.

**Comment by Mikilll94** [comment]
> Essentially a duplicate of #42775.
> 
> Objects in TypeScript are not sealed. Additional properties can exist. The object `{prop2: 'abc', prop1: 123 }` is assignable to your union type, but checking for the existence of `prop1` does **not** mean it's a `string`.

But if have a check for `someVar.prop1` in `if` condition then it means this property exist so i should be able to do i.e. a `console.log`

**Comment by MartinJohns** [comment]
The property existing does not mean you have the type `{ prop1: string }`. Please read the issue I linked.

**Comment by Mikilll94** [comment]
@MartinJohns 

But for `in` operator Typescript does the narrowing. So why it cannot do for property existance check? What's the logic here? In the linked thread I cannot find the answer.

**Comment by MartinJohns** [comment]
The `in` operator is intentionally unsound. This was mentioned in the issue I linked and kindly asked you to read: https://github.com/microsoft/TypeScript/issues/42775#issuecomment-778419557

**Comment by Mikilll94** [comment]
> Essentially a duplicate of #42775.
> 
> Objects in TypeScript are not sealed. Additional properties can exist. The object `{prop2: 'abc', prop1: 123 }` is assignable to your union type, but checking for the existence of `prop1` does **not** mean it's a `string`.

How objects in TypeScript are not sealed?

![image](https://github.com/microsoft/TypeScript/assets/17653648/e5a892de-e005-49ea-90b0-723b2cd8bfb6)

As you can see it's not possible to add additional properties in type.

**Comment by MartinJohns** [comment]
That's the "excess property check", which is more of a linter feature than part of the type system. It only applies to object literals with a known type.

Here's an example of what I meant:

```typescript
type Person = {
    firstName: string;
    lastName: string;
}

const bob = { firstName: 'Bob', lastName: 'Bobsen', age: 37 }

// Object with age property is assignable to Person
const person: Person = bob;

// Or can be passed to a function accepting a Person
function takeHostage(person: Person) {}
takeHostage(bob);
```
[Playground link](https://www.typescriptlang.org/play?#code/C4TwDgpgBAChBOBnA9gOygXigbwFBQKgDMBLJYAOQEMBbCALikWHhNQHMBufQgGyubU6jZqw7cAvrlwBjNMygAjZIsw5iZQbQZQA5ACEVugDRR+W4XsOLEEVCahV2OgMwB2KFNwB6b1ADyigBWEDLAUADuJMAAFo7OUGDwyJDwoFAkiI6IiCTsqFSKvNDAyLAIKKiy8uGplYxwSGhqyorcPn7+8FAyVOiK0GACtgAmUKWOxACuqGEkzVQyMhBgwGzsk42VuEQzc83AVADWEAASyMxOEAAUdWgNFWgAlDhShyfnl87XrU-cQA)

This is working as intended. See also: [FAQ: What is structural typing?](https://github.com/microsoft/TypeScript/wiki/FAQ#what-is-structural-typing)
